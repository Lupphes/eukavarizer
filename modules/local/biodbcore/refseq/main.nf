process BIODBCORE_REFSEQ {
    tag "$taxonomy_id"
    label 'process_low'

    conda "${moduleDir}/environment.yml"
    container "docker.io/luppo/biodbcore:1b8dd90e307a6937b09b9d7088f3b8ec5e8dacc8"

    input:
    val  taxonomy_id
    val  outdir
    path reference_genome

    output:
    path "*_refseq_results.json"         , emit: json
    path "*.{fna.gz,fa.gz}"              , emit: reference_genome, optional: true
    path "assembly_summary_refseq.parquet", emit: refseq_parquet, optional: true
    path "assembly_summary_refseq.txt"   , emit: refseq_summary, optional: true
    path "versions.yml"                  , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${taxonomy_id}"
    def reference_arg = reference_genome ? "--reference_genome ${reference_genome}" : ''
    """
    if [ -n "${reference_genome}" ]; then
        echo "Reference genome already exists at: ${reference_genome} for taxonomy ID: ${taxonomy_id}"
    else
        echo "Downloading reference genome for taxonomy ID: ${taxonomy_id}"
    fi

    biodbcore \\
        --mode refseq \\
        --taxonomy_id ${taxonomy_id} \\
        --outdir . \\
        ${reference_arg} \\
        ${args}

    # Rename output with configurable prefix
    mv refseq_results.json ${prefix}_refseq_results.json

    cat ${prefix}_refseq_results.json

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        biodbcore: \$( biodbcore --version 2>&1 | sed 's/biodbcore //g' )
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${taxonomy_id}"
    """
    # Create mock reference genome if not provided
    if [ -z "${reference_genome}" ]; then
        echo ">mock_chromosome_1" | gzip > ${prefix}_mock_genome.fna.gz
        echo '{"genome_size": 12000000, "genome_size_ungapped": 11500000, "reference_genome": "'${prefix}'_mock_genome.fna.gz"}' > ${prefix}_refseq_results.json
    else
        echo '{"genome_size": 12000000, "genome_size_ungapped": 11500000, "reference_genome": "'${reference_genome}'"}' > ${prefix}_refseq_results.json
    fi

    # Create mock assembly files
    touch assembly_summary_refseq.txt
    touch assembly_summary_refseq.parquet

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        biodbcore: \$( biodbcore --version 2>&1 | sed 's/biodbcore //g' )
    END_VERSIONS
    """
}
