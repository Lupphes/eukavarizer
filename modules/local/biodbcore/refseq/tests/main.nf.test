nextflow_process {

    name "Test Process BIODBCORE_REFSEQ"
    script "../main.nf"
    process "BIODBCORE_REFSEQ"
    tag "biodbcore"
    tag "biodbcore/refseq"
    tag "modules"
    tag "modules_local"

    test("Download reference genome - yeast - stub") {

        options '-stub'

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = 4932  // Saccharomyces cerevisiae
                input[1] = "$outputDir"
                input[2] = []
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() },
                { assert process.out.json },
                { assert process.out.reference_genome },
                { assert process.out.versions }
            )
        }
    }

    test("Use existing reference genome - stub") {

        options '-stub'

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = 9606  // Homo sapiens
                input[1] = "$outputDir"
                input[2] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta')
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() },
                { assert process.out.json },
                { assert process.out.versions }
            )
        }
    }

    test("Invalid taxonomy ID - stub") {

        options '-stub'

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = 99999999  // Invalid taxonomy ID
                input[1] = "$outputDir"
                input[2] = []
                """
            }
        }

        then {
            assertAll(
                { assert process.out.json },
                { assert process.out.versions }
            )
        }
    }
}
