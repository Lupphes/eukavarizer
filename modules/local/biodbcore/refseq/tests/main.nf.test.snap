{
    "Download reference genome - yeast - stub": {
        "content": [
            {
                "0": [
                    "4932_refseq_results.json:md5,6675d18854904d31b7a604c2dfef5880"
                ],
                "1": [
                    "4932_mock_genome.fna.gz:md5,166005cfb5760598d4a29700cdbe38ff"
                ],
                "2": [
                    "assembly_summary_refseq.parquet:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "3": [
                    "assembly_summary_refseq.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "4": [
                    "versions.yml:md5,7042e555b1412d9a01791eeddc2b235b"
                ],
                "json": [
                    "4932_refseq_results.json:md5,6675d18854904d31b7a604c2dfef5880"
                ],
                "reference_genome": [
                    "4932_mock_genome.fna.gz:md5,166005cfb5760598d4a29700cdbe38ff"
                ],
                "refseq_parquet": [
                    "assembly_summary_refseq.parquet:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "refseq_summary": [
                    "assembly_summary_refseq.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "versions": [
                    "versions.yml:md5,7042e555b1412d9a01791eeddc2b235b"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-23T01:19:14.954600511"
    },
    "Use existing reference genome - stub": {
        "content": [
            {
                "0": [
                    "9606_refseq_results.json:md5,aecc507a4a9772d0538fc3f9159295b8"
                ],
                "1": [
                    
                ],
                "2": [
                    "assembly_summary_refseq.parquet:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "3": [
                    "assembly_summary_refseq.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "4": [
                    "versions.yml:md5,7042e555b1412d9a01791eeddc2b235b"
                ],
                "json": [
                    "9606_refseq_results.json:md5,aecc507a4a9772d0538fc3f9159295b8"
                ],
                "reference_genome": [
                    
                ],
                "refseq_parquet": [
                    "assembly_summary_refseq.parquet:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "refseq_summary": [
                    "assembly_summary_refseq.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "versions": [
                    "versions.yml:md5,7042e555b1412d9a01791eeddc2b235b"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-23T01:19:34.845011462"
    }
}